tool/tsh/common/tsh.go

```diff
@@ -4143,6 +4145,51 @@ func refuseArgs(command string, args []string) error {
 	return nil
 }
 
+// onFlatten reads an identity file and flattens it into a tsh profile on disk.
+func onFlatten(cf *CLIConf) error {
+	// Save the identity file path for later
```

Comment:
LMK what you think about the patch below, which handles a few suggestions I wanted to make:
- Prevent `tsh flatten` from overwriting existing tsh profiles. IMO, we should try to assimilate the key into existing profiles if applicable, only overwriting overlapping keys.
- Reuse code in the `identityfile` package instead of copying some of it here
- Reuse the code we use to save profile information from proxy ping when necessary. 
- Make sure we are parsing the `cf` input before using it.
  - This will do things like turn `--proxy=root.example.com` into `--proxy=root.example.com:3080` if it should. I think this may also be where we handle the `TELEPORT_HOME` envar.
  - The most efficient way to do this is to just call `loadClientConfigFromCLIConf` and use the resulting config. This will also initialize the file store and do other things we may want, without going ahead and creating a full client.
- Report the login status after `tsh flatten`.

<details>
<summary>Patch</summary>
diff --git a/api/profile/profile.go b/api/profile/profile.go
index 78831c8805..932afed613 100644
--- a/api/profile/profile.go
+++ b/api/profile/profile.go
@@ -106,6 +106,10 @@ type Profile struct {
 
 	// PIVSlot is a specific piv slot that Teleport clients should use for hardware key support.
 	PIVSlot keys.PIVSlot `yaml:"piv_slot"`
+
+	// MissingClusterDetails means this profile was created with limited cluster details.
+	// Missing cluster details should be loaded into the profile by pinging the proxy.
+	MissingClusterDetails bool
 }
 
 // Copy returns a shallow copy of p, or nil if p is nil.
diff --git a/lib/client/identityfile/identity.go b/lib/client/identityfile/identity.go
index 0b10e6be95..1070abfc60 100644
--- a/lib/client/identityfile/identity.go
+++ b/lib/client/identityfile/identity.go
@@ -811,35 +811,51 @@ func KeyFromIdentityFile(identityPath, proxyHost, clusterName string) (*client.K
 // Additionally, the [clusterName] argument can ve used to target a leaf cluster
 // rather than the default root cluster.
 func NewClientStoreFromIdentityFile(identityFile, proxyAddr, clusterName string) (*client.Store, error) {
+	clientStore := client.NewMemClientStore()
+	if err := LoadIdentityFileIntoClientStore(clientStore, identityFile, proxyAddr, clusterName); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return clientStore, nil
+}
+
+// LoadIdentityFileIntoClientStore loads the identityFile from the given path
+// into the given client store, assimilating it with other keys in the store.
+func LoadIdentityFileIntoClientStore(store *client.Store, identityFile, proxyAddr, clusterName string) error {
 	if proxyAddr == "" {
-		return nil, trace.BadParameter("missing a Proxy address when loading an Identity File.")
+		return trace.BadParameter("missing a Proxy address when loading an Identity File.")
 	}
 	proxyHost, err := utils.Host(proxyAddr)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return trace.Wrap(err)
 	}
 
 	key, err := KeyFromIdentityFile(identityFile, proxyHost, clusterName)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return trace.Wrap(err)
 	}
 
-	// Preload the client key from the agent.
-	clientStore := client.NewMemClientStore()
-	if err := clientStore.AddKey(key); err != nil {
-		return nil, trace.Wrap(err)
+	// Load the client key from the agent.
+	if err := store.AddKey(key); err != nil {
+		return trace.Wrap(err)
 	}
 
-	// Save temporary profile into the key store.
-	profile := &profile.Profile{
-		WebProxyAddr:     proxyAddr,
-		SiteName:         key.ClusterName,
-		Username:         key.Username,
-		PrivateKeyPolicy: key.PrivateKey.GetPrivateKeyPolicy(),
-	}
-	if err := clientStore.SaveProfile(profile, true); err != nil {
-		return nil, trace.Wrap(err)
+	// If the client store is not initialized (this is the first key added),
+	// then the profile will be missing. Add a temporary profile with necessary
+	// information. This profile should be filled in with information from the
+	// proxy by the caller.
+	if _, err := store.GetProfile(proxyAddr); trace.IsNotFound(err) {
+		profile := &profile.Profile{
+			MissingClusterDetails: true,
+			WebProxyAddr:          proxyAddr,
+			SiteName:              key.ClusterName,
+			Username:              key.Username,
+			PrivateKeyPolicy:      key.PrivateKey.GetPrivateKeyPolicy(),
+		}
+		if err := store.SaveProfile(profile, true); err != nil {
+			return trace.Wrap(err)
+		}
 	}
 
-	return clientStore, nil
+	return nil
 }
diff --git a/tool/tsh/common/tsh.go b/tool/tsh/common/tsh.go
index d38f66f29b..ccf5b7fb7f 100644
--- a/tool/tsh/common/tsh.go
+++ b/tool/tsh/common/tsh.go
@@ -3560,15 +3560,16 @@ func makeClientForProxy(cf *CLIConf, proxy string) (*client.TeleportClient, erro
 
 	// If we are missing client profile information, ping the webproxy
 	// for proxy info and load it into the client config.
-	if profileError != nil || cf.IdentityFileIn != "" {
+	if profileError != nil || profile.MissingClusterDetails {
 		log.Debug("Pinging the proxy to fetch listening addresses for non-web ports.")
 		_, err := tc.Ping(cf.Context)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
 
-		// Identityfile uses a placeholder profile. Save missing profile info.
-		if cf.IdentityFileIn != "" {
+		// This is a placeholder profile created from limited cluster details.
+		// Save missing cluster details gathererd during Ping.
+		if profileError == nil && profile.MissingClusterDetails {
 			if err := tc.SaveProfile(true); err != nil {
 				return nil, trace.Wrap(err)
 			}
@@ -4156,38 +4157,26 @@ func onFlatten(cf *CLIConf) error {
 		return trace.BadParameter("--proxy must be specified")
 	}
 
-	// Making a client validates the connection and builds a valid profile for us
-	tc, err := makeClientForProxy(cf, cf.Proxy)
-	if err != nil {
-		return trace.Wrap(err, "making client for proxy")
-	}
-	profile := tc.Profile()
-
-	// We clear the identity flag and create a new client store.
-	// Because the identity flag is unset, it will be backed by the filesystem (in ~/.tsh or TELEPORT_HOME).
+	// We clear the identity flag so that the client store will be backed
+	// by the filesystem instead (in ~/.tsh or TELEPORT_HOME).
 	cf.IdentityFileIn = ""
-	store, err := initClientStore(cf, cf.Proxy)
-	if err != nil {
-		return trace.Wrap(err, "initializing client store")
-	}
 
-	// Now we extract the key from the identity file and load it into our new store.
-	key, err := identityfile.KeyFromIdentityFile(identityFile, cf.Proxy, cf.SiteName)
+	// Load client config as normal to parse client inputs and add defaults.
+	c, err := loadClientConfigFromCLIConf(cf, cf.Proxy)
 	if err != nil {
-		return trace.Wrap(err, "loading key from identity file")
+		return trace.Wrap(err)
 	}
 
-	if err = store.AddKey(key); err != nil {
-		return trace.Wrap(err, "adding key into the client store")
+	// Load the identity file key and partial profile into the client store.
+	if err := identityfile.LoadIdentityFileIntoClientStore(c.ClientStore, identityFile, c.WebProxyAddr, c.SiteName); err != nil {
+		return trace.Wrap(err)
 	}
 
-	// Finally we also save the profile built by the client into our new store.
-	if err := store.SaveProfile(profile, true); err != nil {
-		return trace.Wrap(err, "saving profile")
-	}
+	fmt.Printf("Successfully flattened Identity file %q into a tsh profile.\n", identityFile)
 
-	fmt.Printf("Identity file %q flattented into a tsh profile.", identityFile)
-	return nil
+	// onStatus will ping the proxy to fill in cluster profile information missing in the
+	// client store, then print the login status.
+	return trace.Wrap(onStatus(cf))
 }
 
 // onShow reads an identity file (a public SSH key or a cert) and dumps it to stdout
</details>